File: Hard/1293. Shortest Path in a Grid with Obstacles Elimination/ShortestPathInAGridWithObstaclesElimination.java
/*
LeetCode Hard: 1293. Shortest Path in a Grid with Obstacles Elimination
https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/

Check ShortestPathInAGridWithObstaclesElimination.cpp for brief solution explanation.
*/
class Solution {
    public boolean bounds(int r, int c, int M, int N){
        return 0<=r&&r<M&&0<=c&&c<N;
    }
    public int shortestPath(int[][] grid, int k) {
        int M = grid.length, N = grid[0].length;
        int[][][]dis = new int[M][N][k+1];
        boolean[][][]vis = new boolean[M][N][k+1];
        int[][]moves = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}}; //Up, Down, Left, Right
        Queue<int[]>q = new LinkedList<int[]>();
        dis[0][0][0] = 0;
        vis[0][0][0] = true;
        q.add(new int[]{0, 0, 0});
        while(!q.isEmpty()){
            int[]cur = q.poll();
            int r = cur[0], c = cur[1], z = cur[2];
            if(r==M-1&&c==N-1){
                return dis[r][c][z];
            }
            for(int i = 0;i<4;i++){
                int newR = r+moves[i][0], newC = c+moves[i][1];
                if(bounds(newR, newC, M, N)){
                    if(grid[newR][newC]==1&&z<k&&!vis[newR][newC][z+1]){
                        dis[newR][newC][z+1] = dis[r][c][z]+1;
                        vis[newR][newC][z+1] = true;
                        q.add(new int[]{newR, newC, z+1});
                    }
                    if(grid[newR][newC]==0&&!vis[newR][newC][z]){
                        dis[newR][newC][z] = dis[r][c][z]+1;
                        vis[newR][newC][z] = true;
                        q.add(new int[]{newR, newC, z});
                    }
                }
            }
        }
        return -1;
    }
}

File: Hard/1293. Shortest Path in a Grid with Obstacles Elimination/ShortestPathInAGridWithObstaclesElimination.py
'''
LeetCode Hard: 1293. Shortest Path in a Grid with Obstacles Elimination
https://leetcode.com/problems/shortest-path-in-a-grid-with-obstacles-elimination/

Check ShortestPathInAGridWithObstaclesElimination.cpp for brief solution explanation.
'''
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        M, N = len(grid), len(grid[0])
        def bounds(r, c) -> bool:
            return 0<=r<M and 0<=c<N
        dis = [[[0 for i in range(k+1)]for i in range(N)]for i in range(M)]
        vis = [[[False for i in range(k+1)]for i in range(N)]for i in range(M)]
        dis[0][0][0] = 0
        vis[0][0][0] = True
        moves = [[0, -1], [0, 1], [-1, 0], [1, 0]] #Up, Down, Left, Right
        q = deque([[0, 0, 0]])
        while len(q)>0:
            cur = q.popleft()
            r, c, z = cur[0], cur[1], cur[2]
            if r==M-1 and c==N-1:
                return dis[r][c][z]
            for i in range(4):
                newR, newC = r+moves[i][0], c+moves[i][1]
                if bounds(newR, newC):
                    if grid[newR][newC]==1 and z<k and not vis[newR][newC][z+1]:
                        dis[newR][newC][z+1] = dis[r][c][z]+1
                        vis[newR][newC][z+1] = True
                        q.append([newR, newC, z+1])
                    if grid[newR][newC]==0 and not vis[newR][newC][z]:
                        dis[newR][newC][z] = dis[r][c][z]+1
                        vis[newR][newC][z] = True
                        q.append([newR, newC, z])
        return -1